compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/MoveCompanionObjectFieldsLowering.kt

```diff
@@ -66,26 +67,25 @@ private class MoveCompanionObjectFieldsLowering(val context: CommonBackendContex
         val companion = irClass.declarations.find {
             it is IrClass && it.isCompanion
         } as IrClass? ?: return
-        if ((irClass.isInterface || irClass.isAnnotationClass) && !companion.allFieldsAreJvmField()) return
-        companion.declarations.forEach {
-            when (it) {
-                is IrProperty -> {
-                    val newField = movePropertyFieldToStaticParent(it, companion, irClass, fieldReplacementMap)
-                    if (newField != null) irClass.declarations.add(newField)
-                }
-                is IrAnonymousInitializer -> {
-                    val newInitializer = moveAnonymousInitializerToStaticParent(it, companion, irClass)
-                    irClass.declarations.add(newInitializer)
-                }
-                else -> Unit
+        companion.declarations.mapNotNullTo(irClass.declarations) {
+            when {
+                it is IrProperty && shouldMoveProperty(irClass, it) ->
+                    movePropertyFieldToStaticParent(it, companion, irClass, fieldReplacementMap)
+                it is IrAnonymousInitializer ->
+                    moveAnonymousInitializerToStaticParent(it, companion, irClass)
+                else -> null
             }
         }
         companion.declarations.removeAll { it is IrAnonymousInitializer }
     }
 
-    private fun IrClass.allFieldsAreJvmField() =
-        declarations.filterIsInstance<IrProperty>()
-            .mapNotNull { it.backingField }.all { it.hasAnnotation(JVM_FIELD_ANNOTATION_FQ_NAME) }
+    // Properties in companions are always copied into their declaration parents except for interfaces, which can only host
```

Comment:
@ting-yuan Seems it's a bug in documentation around interface companion properties.
Actually by default all properties in interface companion are not moved to outer interface. 'Const' properties just copied to it (so there are 2 representation of them).
If interface companion have all fields marked as @JvmFiled then only in such case all fields are moved.

Reply:
@max-kammerer So, for example, if a companion object in an interface has a field marked with @JvmField and another field without, both of them shouldn't be moved nor copied into the outer interface. Is that correct? If that is the case, how do Java refer to the one with @JvmField?

I uploaded a patch implementing that. Please take a look if this is the desired behavior.

Reply:
@ting-yuan Yes it's correct, @JvmField in such cases is prohibeted (it's not applicable by default in interface companion). Doc team should update documentation